File: Numeros Pseudo-Aleatorios/Permuted Congruential Generator/pcg.java
/**
 * PRNG .......: Permuted Congruential Generator (PCG)
 * Ano ........: 2014
 * Autor ......: Dr. M.E. O'Neill
 */

import java.util.Date;

public class Main {

    private static int rotr32(int x, int r) {
        return (x >>> r) | (x << (-r & 31));
    }

    private static int pcg32(long seed) {
        long x = seed;
        int rotation = (int) (x >>> 59);
        x = x * 6364136223846793005L + 1442695040888963407L;
        x ^= x >>> 18;
        x = x + seed;
        return rotr32((int) (((x >>> 27) + rotation) ^ (x >>> 32)), rotation) & 0x7FFFFFFF;
    }

    private static int getTimer() {
        long millis = new Date().getTime();
        return (int) (millis & 0xFFFFFFFFL);
    }

    public static void main(String[] args) {
		
        System.out.println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)");
        System.out.println("PSEUDORANDOM NUMBER GENERATOR (PRNG)");
        System.out.println("PADRAO: Permuted Congruential Generator (PCG)");
        System.out.println("=============================================");
        System.out.println();

        long Xn1 = 1234;
        long Xn2 = getTimer();

        System.out.println("      Semente: 1234       Semente: TIME");
        System.out.println("      -------------       -------------");

        for (int i = 1; i <= 10; i++) {
            int random1 = pcg32(Xn1);
            int random2 = pcg32(Xn2);
            System.out.printf("%19d", random1);
            System.out.printf("%20d%n", random2);
            Xn1 = random1;
            Xn2 = random2;
        }
        
    }
    
}

File: Numeros Pseudo-Aleatorios/Permuted Congruential Generator/pcg.scala
/*
 * PRNG .......: Permuted Congruential Generator (PCG)
 * Ano ........: 2014
 * Autor ......: Dr. M.E. O'Neill
 */

object Main {
  def main(args: Array[String]): Unit = {
    import java.time.{Instant, Duration}

    def rotr32(x: Long, r: Int): Long = (x >>> r) | (x << (32 - r))

    def pcg32(seed: Array[Long]): Int = {
      var x = seed(0)
      val rotation = (x >>> 59).toInt
      seed(0) = x * 6364136223846793005L + 1442695040888963407L
      x ^= x >>> 18
      seed(0) += x
      ((rotr32(((x >>> 27) + rotation) ^ (seed(0) >>> 32), rotation) >>> 1) & 0x7FFFFFFF).toInt
    }

    def getTimer: Long = Instant.now().toEpochMilli

    val Xn1 = Array[Long](1234)
    val Xn2 = Array[Long](getTimer)

    println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
    println("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
    println("PADRAO: Permuted Congruential Generator (PCG)")
    println("=============================================")
    println()

    println("      Semente: 1234       Semente: TIME")
    println("      -------------       -------------")

    for (i <- 1 to 10) {
      print(f"${pcg32(Xn1)}%19d")
      println(f"${pcg32(Xn2)}%20d")
    }
  }
}
